src/writer/summarized.rs

```diff
@@ -0,0 +1,196 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! [`Writer`]-wrapper for collecting a summary of execution.
+
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use derive_more::Deref;
+
+use crate::{event, OutputtedWriter, World, Writer};
+
+/// [`Writer`] for collecting summary: number of features, scenarios and steps.
+///
+/// Wrapper for a [`Writer`] implementation for outputting a summary (number of
+/// features, scenarios, steps and parsing errors) of execution.
+#[derive(Debug, Deref)]
+pub struct Summarized<Writer> {
+    #[deref]
+    /// Inner [`Writer`].
+    pub writer: Writer,
+
+    /// Number of started [`Feature`]s.
+    ///
+    /// [`Feature`]: gherkin::Feature
+    pub features: usize,
+
+    /// Number of started [`Rule`]s.
+    ///
+    /// [`Rule`]: gherkin::Rule
+    pub rules: usize,
+
+    /// Number of started [`Scenario`]s.
+    ///
+    /// [`Scenario`]: gherkin::Scenario
+    pub scenarios: usize,
+
+    /// Number of [`Parser`] errors.
+    ///
+    /// [`Parser`]: crate::Parser
+    pub parsing_errors: usize,
+
+    /// [`Step`]s [`Stats`]
+    ///
+    /// [`Step`]: gherkin::Step
+    pub steps: Stats,
+}
+
+/// [`Step`]s statistics.
+///
+/// [`Step`]: gherkin::Step
+#[derive(Clone, Copy, Debug)]
+pub struct Stats {
+    /// Number of passed [`Step`]s.
+    ///
+    /// [`Step`]: gherkin::Step
+    pub passed: usize,
+
+    /// Number of skipped [`Step`]s.
+    ///
+    /// [`Step`]: gherkin::Step
+    pub skipped: usize,
+
+    /// Number of failed [`Step`]s.
+    ///
+    /// [`Step`]: gherkin::Step
+    pub failed: usize,
+}
+
+#[async_trait(?Send)]
+impl<W, Wr> Writer<W> for Summarized<Wr>
+where
+    W: World,
+    Wr: for<'val> OutputtedWriter<'val, W, String>,
+{
+    async fn handle_event(&mut self, ev: event::Cucumber<W>) {
+        use event::{Cucumber, Feature, Rule};
+
+        let mut finished = false;
+        match &ev {
+            Cucumber::ParsingError(_) => self.parsing_errors += 1,
+            Cucumber::Feature(_, ev) => match ev {
+                Feature::Started => self.features += 1,
+                Feature::Rule(_, Rule::Started) => {
+                    self.rules += 1;
+                }
+                Feature::Rule(_, Rule::Scenario(_, ev))
+                | Feature::Scenario(_, ev) => self.handle_scenario(ev),
+                Feature::Finished | Feature::Rule(..) => {}
+            },
+            Cucumber::Finished => finished = true,
+            Cucumber::Started => {}
+        };
+
+        self.writer.handle_event(ev).await;
+
+        if finished {
+            let summary = format!(
+                "[Summary]\n\
+                 {} features\n\
+                 {} rules\n\
+                 {} scenarios\n\
+                 {} steps ({} passed, {} skipped, {} failed)\n\
+                 {} parsing errors",
```

Comment:
@ilslv this still will be too bad with `JsonWriter` or something like that. Whe should abstract more specificaly here to not hardcode any possible format, but rather to give the underlying `Writer` an ability to format the structured data as it wants.

Reply:
@tyranron we already can describe desired format

```rust
//                                 ˅ Generalised by `Value`
pub trait Arbitrary<'val, World, Value: 'val>: Writer<World> {
    /// Writes `val` to the [`Writer`]'s output.
    async fn write(&mut self, val: Value)
    where
        'val: 'async_trait;
}
```

```rust
#[async_trait(?Send)]
impl<W, Wr> Writer<W> for Summarized<Wr>
where
    W: World,
    //                                          So we can work only with `ArbitraryWriter` 
    //                                        ˅ which expects `String` as input 
    Wr: for<'val> ArbitraryWriter<'val, W, String>,
{
    async fn handle_event(&mut self, ev: event::Cucumber<W>) {
        // ...
        if finished {
            let summary = format!(
                "[Summary]\n\
                 {} features\n\
                 {} rules\n\
                 {} scenarios\n\
                 {} steps ({} passed, {} skipped, {} failed)\n\
                 {} parsing errors",
                self.features,
                self.rules,
                self.scenarios,
                self.steps.passed + self.steps.skipped + self.steps.failed,
                self.steps.passed,
                self.steps.skipped,
                self.steps.failed,
                self.parsing_errors,
            );
            self.writer.write(summary).await;
        }
    }
}
```

Reply:
@ilslv oh, here is the tweak! Thanks, quite neat! Can you/describe mention that in documentation with an example, so this will be clear for future contributors?